include/zephyr/device.h

```diff
@@ -906,8 +904,9 @@ static inline bool z_impl_device_is_ready(const struct device *dev)
 #define Z_DEVICE_BASE_DEFINE(node_id, dev_id, name, pm, data, config, level,   \
 			     prio, api, state, handles)                        \
 	COND_CODE_1(DT_NODE_EXISTS(node_id), (), (static))                     \
-	const Z_DECL_ALIGN(struct device) DEVICE_NAME_GET(                     \
-		dev_id) Z_DEVICE_SECTION(level, prio) __used =                 \
+	const STRUCT_SECTION_ITERABLE_NAMED(device,                            \
+		Z_DEVICE_SECTION_NAME(level, prio),                            \
```

Comment:
Does these actually need to be sorted by level and priority?

Reply:
Yes, PM subsystem relies on devices list being sorted. However, what exists in PM subsys for devices, except runtime PM it's pretty useless in many (if not all) applications.

Reply:
Is the sorting alphabetically? Could this be an issue for single digit priorities?

```
POST_KERNEL1
POST_KERNEL11
POST_KERNEL12
...
POST_KERNEL2
POST_KERNEL21
POST_KERNEL22
...
```

Reply:
@pdgendt you're right, let me try to find a solution.

Reply:
@pdgendt should be fixed now

Reply:
Nice. Can you add `ITERABLE_SECTION_RAM_NUMERIC` for users that would like the `RAM` variant?

Reply:
@pdgendt actually, current code won't work as levels need to be manually sorted... PRE_KERNEL_1 < APPLICATION. Let me add some iterable helpers that allow to define a section with "subsections".